File: 2196-create-binary-tree-from-descriptions/2196-create-binary-tree-from-descriptions.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        root_candidate = set()
        known_child = set()
        dict_description = dict()
        for x in descriptions:
            if x[0] not in dict_description:
                dict_description[x[0]] = [0, 0]
            
            if x[2]==1:
                dict_description[x[0]][0] = x[1]
            else:
                dict_description[x[0]][1] = x[1]

            if x[1] in root_candidate:
                root_candidate.remove(x[1])
                known_child.add(x[1])
            elif x[1] not in known_child:
                known_child.add(x[1])
            if x[0] not in root_candidate and x[0] not in known_child:
                root_candidate.add(x[0])
        
        # set is not subscriptable, so this is the only way to get
        for i in root_candidate:
            root = TreeNode(i, None, None)

        def dfs(root):
            if not root or root.val not in dict_description:
                return

            childs = dict_description[root.val]
            if childs[0]!=0:
                root.left = TreeNode(childs[0], None, None)
                dfs(root.left)
            if childs[1]!=0:
                root.right = TreeNode(childs[1], None, None)
                dfs(root.right)

        dfs(root)
        return root

File: 2196-create-binary-tree-from-descriptions/2196-create-binary-tree-from-descriptions.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* createBinaryTree(vector<vector<int>>& descriptions) {
        // use set to store distinct values
        unordered_set<int> childs, parents;
        // store the parent, child relation in the map
        unordered_map<int, vector<pair<int, int>>> map;

        int n_descriptions = descriptions.size();
        for (int i=0; i<n_descriptions; i++)
        {
            int parent = descriptions[i][0], child = descriptions[i][1], left = descriptions[i][2];
            parents.insert(parent);
            childs.insert(child);
            map[parent].push_back(make_pair(child, left));
        }

        // check all the values in parents and remove the ones in child to find the root
        for (auto curr = parents.begin(); curr != parents.end();)
        {
            if (childs.find(*curr)!=childs.end())
            {
                // move to the next element in parents
                curr = parents.erase(curr);
            }
            else
            {
                curr++;
            }
        }

        // initialize the root
        TreeNode * root = new TreeNode(*parents.begin());
        queue<TreeNode *> queue;
        queue.push(root);

        // use bfs to construct the tree
        while (!queue.empty())
        {
            TreeNode * parent = queue.front();
            queue.pop();

            // find all the existing child for the parents
            for (auto& childInfo : map[parent->val])
            {
                int childValue = childInfo.first, left = childInfo.second;
                TreeNode * node = new TreeNode(childValue);
                queue.push(node);

                if (left==1)
                {
                    parent->left = node;
                }
                else
                {
                    parent->right = node;
                }
            }
        }

        return root;
    }
};